File: c++/06_pgm_avion/src/co/edu/campusucc/poo/Avion.cpp
#include <iostream>
#include <string>

using namespace std;

class Avion {
public:
    Avion() {}

    // Properties
    string marca;
    string tipo;
    int velocidad;

    string toString() {
        return "Clase Abstracta--> Avion: " + marca + " " + tipo + " ";
    }

    // Abstract Behavior
    virtual void startingEngine() = 0;

    virtual void stopingEngine() = 0;

    // Definir comportamiento
    void speedUp() {
        string speedUpString = "SpeedUp in progres: ";
        for (int i = 0; i < velocidad; i++) {
            speedUpString += "🛫";
        }

        for (int i = 0; i < 10; i++) {
            speedUpString += "🛫";
            cout << speedUpString << endl;
            try {
                this_thread::sleep_for(chrono::milliseconds(1000 - (i * 50)));
            } catch (const exception& e) {
                cerr << "⛔:" << e.what() << endl;
            }
        }
        addSpeed(10);

    }

protected:
    int addSpeed(int velocidad) {
        this->velocidad += velocidad;
        return velocidad;
    }

    int downSpeed(int velocidad) {
        this->velocidad -= velocidad;
        return velocidad;
    }

public:
    void currentSpeed() {
        cout << "Current Speed 🛫: " << velocidad << endl;
    }

    void landing() {
        for (int i = 0; i < 5; i++) {
            cout << "🛬🛬 Landing in Progress 🚷🚷" << endl;
            try {
                this_thread::sleep_for(chrono::milliseconds(1000 - (i * 50)));
            } catch (const exception& e) {
                cerr << "⛔:" << e.what() << endl;
            }
        }

    }

    void shootMissiles() {
        for (int i = 0; i < 5; i++) {
            cout << "🚀🚀 Missiles ready to fire 🔥🔥🔥" << endl;
            try {
                this_thread::sleep_for(chrono::milliseconds(1000 - (i * 50)));
            } catch (const exception& e) {
                cerr << "⛔:" << e.what() << endl;
            }
        }
    }

    void irrigation() {
        for (int i = 0; i < 5; i++) {
            cout << "🛩️🛩️ irrigation in process 💧💧💧" << endl;
            try {
                this_thread::sleep_for(chrono::milliseconds(1000 - (i * 50)));
            } catch (const exception& e) {
                cerr << "⛔:" << e.what() << endl;
            }
        }
    }

    void descent() {
        string speedUpString = "";
        for (int i = 0; i < velocidad; i++) {
            speedUpString += "🛬";
        }
        for (int i = 0; i < 10; i++) {
            speedUpString = speedUpString.substr(0, speedUpString.length() - 1);
            cout << "🛬🛬Descent in Progress🛬🛬: " << speedUpString << endl;
            try {
                this_thread::sleep_for(chrono::milliseconds(1000 - (veloc

File: ruby/pgms_ruby/09-pgmAvionInterface/Avion.rb
require_relative 'IAvionStandar'
require 'time'

class Avion
  include IAvionStandar
  attr_accessor :marca, :tipo, :velocidad

  def initialize
    @marca = nil
    @tipo = nil
    @velocidad = 0
  end

  def to_s
    "Clase Abstracta--> Avion: #{@marca} #{@tipo}"
  end

  # Abstract Behavior
  def startingEngine
    raise NotImplementedError, "#{self.class} no ha implementado el método '#{__method__}'"
  end

  def stopingEngine
    raise NotImplementedError, "#{self.class} no ha implementado el método '#{__method__}'"
  end

  # Definir comportamiento
  def speedUp
    speedUpString = "SpeedUp in progres: "

    @velocidad.times do
      speedUpString += "🛫"
    end

    10.times do |i|
      speedUpString += "🛫"
      puts speedUpString

      begin
        sleep(1 - (i * 0.05))
      rescue Exception => e
        puts "⛔:#{e.message}"
      end
    end

    addSpeed(10)
  end

  def addSpeed(velocidad)
    @velocidad += velocidad
    @velocidad
  end

  def downSpeed(velocidad)
    @velocidad -= velocidad
    @velocidad
  end

  def currentSpeed
    puts "Current Speed 🛫: #{@velocidad}"
  end

  def landing
    5.times do
      puts "🛬🛬 Landing in Progress 🚷🚷"

      begin
        sleep(1)
      rescue Exception => e
        puts "⛔:#{e.message}"
      end
    end
  end

  def descent
    speedUpString = ""

    @velocidad.times do
      speedUpString += "🛬"
    end

    10.times do |i|
      speedUpString = speedUpString[0...-1]
      puts "🛬🛬Descent in Progress🛬🛬: #{speedUpString}"

      begin
        sleep(1 - (@velocidad + i * 0.05))
      rescue Exception => e
        puts "⛔:#{e.message}"
      end
    end

    downSpeed(10)
  end

  # Getters and Setters
  def chargeFuel(litros)
    puts "⛽ Charging Fuel: #{litros} 💧💧💧"
  end
end